File: nim_datalog/query_edn.nim
## query_edn.nim — Datalog EDN surface → DatalogIR.
##
## Parses the Datomic-style query vector into the IR the planner already
## consumes, bypassing the SQL AST entirely:
##
##   [:find  ?name ?price
##    :in    $ ?email
##    :where [?e :person/name ?name]
##           [?e :person/email ?email]
##           [?e :fin/price ?price]
##           [(> ?price 5)]]
##
## Grammar (subset — docs/datalog-reference.md):
##   query    := [:find var+ (:in $ var+)? :where clause+ (:history)?]
##   clause   := pattern | predicate | (or pred+)
##   pattern  := [e-slot attr v-slot]   e: var|_ ; attr: :ns/name ; v: var|_|const
##   predicate:= [(op var value)]       op: > < >= <= = !=
##   or       := (or pred pred+)        same var in every pred
##   var      := ?name   (_ = blank)
##   value    := string|int|float|bool|:kw|?param
##
## `?x` vars in :in bind positionally to query args (param 1..N); pattern
## slots referencing them become dsConst(bvParam).  Attribute keywords
## normalize to the canonical slash form (:person/name → person/name).

import std/[tables, strutils, options, sequtils]
import edn
import scheme  # SExpr — the args passed by the caller are SExprs
import datalog_ast

type
  DatalogSyntaxError* = object of CatchableError

const RangeOps = [">", "<", ">=", "<=", "=", "!="]

proc isVar(e: SExpr): bool = e.kind == sSymbol and e.symval.startsWith("?")

proc varName(e: SExpr): string =
  ## ?name → "name" (the IR stores var names without the ? prefix)
  e.symval[1 ..^ 1]

proc isBlank(e: SExpr): bool =
  e.kind == sSymbol and e.symval == "_"

proc isKeyword(e: SExpr): bool = e.kind == sKeyword

proc parseValue(e: SExpr; params: Table[string, uint32]): BoundValue =
  ## A constant or param in a predicate/pattern value position.
  if isVar(e):
    let vn = varName(e)
    if vn in params: return newBoundParam(params[vn])
    raise newException(DatalogSyntaxError,
      "datalog: unbound var ?" & vn & " in value position (declare it in :in)")
  case e.kind
  of sStr: newBoundStr(e.sval)
  of sInt: newBoundInt(e.ival)
  of sFloat: newBoundFloat(e.fval)
  of sBool: newBoundBool(e.bval)
  of sKeyword: newBoundStr(":" & e.kwval)  # enum-style keyword value
  else:
    raise newException(DatalogSyntaxError,
      "datalog: unsupported value: " & $e)

proc parsePattern(p: SExpr; ir: var DatalogIR; params: Table[string, uint32]) =
  ## [e attr v] with vars, blanks and constants.
  if p.kind != sList or p.items.len != 3:
    raise newException(DatalogSyntaxError,
      "datalog: pattern must be a 3-element vector [e attr v], got " & $p)
  var dp = DatalogPattern(t: slotMissing(), added: slotMissing())

  # e slot: var, _ or int eid
  let eSlot = p.items[0]
  if isVar(eSlot): dp.e = slotVar(varName(eSlot))
  elif isBlank(eSlot): dp.e = slotMissing()
  elif eSlot.kind == sInt: dp.e = slotConst(newBoundInt(eSlot.ival))
  else:
    raise newException(DatalogSyntaxError,
      "datalog: e slot must be a var, _ or eid, got " & $eSlot)

  # a slot: keyword (required — no wildcard patterns in v1)
  let aSlot = p.items[1]
  if not isKeyword(aSlot):
    raise newException(DatalogSyntaxError,
      "datalog: attr slot must be a keyword like :person/name, got " & $aSlot)
  # SQL-style surface (:dl.name) normalizes to canonical slash (dl/name)
  let attr = if "." in aSlot.kwval: aSlot.kwval.replace(".", "/") else: aSlot.kwval
  if "/" notin attr:
    raise newException(DatalogSyntaxError,
      "datalog: attr keyword must be namespaced: :" & aSlot.kwval)
  dp.a = slotConst(newBoundAttr(attr))

  # v slot: var (or :in param), _ or constant
  let vSlot = p.items[2]
  if isVar(vSlot):
    let vn = varName(vSlot)
    if vn in params: dp.v = slotConst(newBoundParam(params[vn]))
    else: dp.v = slotVar(vn)
  elif isBlank(vSlot): dp.v = slotMissing()
  else: dp.v = slotConst(parseValue(vSlot, params))

  ir.patterns.add dp

proc isOpSym(e: SExpr): bool =
  e.kind == sSymbol and e.symval in RangeOps

proc predKey(opExpr: SExpr): string =
  if opExpr.kind != sKeyword:
    raise newException(DatalogSyntaxError,
      "datalog: predicate op must be a keyword like :>, got " & $opExpr)
  let op = opExpr.kwval
  if op notin RangeOps:
    raise newException(DatalogSyntaxError,
      "datalog: unsupported predicate op :" & op &
      " (supported: > < >= <= = !=)")
  op

proc parsePredicate(p: SExpr; ir: var DatalogIR;
                    params: Table[string, uint32]; branch: int) =
  ## [(op ?var value)] — adds a range condition on the var.
  if p.kind != sList or p.items.len != 3:
    raise newException(DatalogSyntaxError,
      "datalog: predicate must be [(op var value)], got " & $p)
  let op = if isOpSym(p.items[0]): p.items[0].symval else: predKey(p.items[0])
  if not isVar(p.items[1]):
    raise newException(DatalogSyntaxError,
      "datalog: predicate var must be ?var, got " & $p.items[1])
  let vn = varName(p.items[1])
  let bv = parseValue(p.items[2], params)
  if vn notin ir.rangeBounds: ir.rangeBounds[vn] = @[]
  while ir.rangeBounds[vn].len <= branch:
    ir.rangeBounds[vn].add(@[])
  ir.rangeBounds[vn][branch].add((op, bv))

proc parseWhereElement(p0: SExpr; ir: var DatalogIR;
                       params: Table[string, uint32]; branch: int = 0) =
  ## Route one :where element.  The EDN reader nests predicate vectors:
  ## [(> ?v 5)] arrives as [[> ?v 5]] with the op as a symbol atom (not a
  ## keyword).  Unwrap one level, then route: or-form → branches, predicate
  ## → range, else → pattern.
  var p = p0
  if p.kind == sList and p.items.len == 1 and p.items[0].kind == sList:
    let inner = p.items[0]
    let innerIsRoute = inner.items.len > 0 and (
      isOpSym(inner.items[0]) or
      (inner.items[0].kind == sSymbol and inner.items[0].symval == "or") or
      (inner.items[0].kind == sKeyword and
       (inner.items[0].kwval == "or" or inner.items[0].kwval in RangeOps)))
    if innerIsRoute:
      p = inner
  let headIsOr = p.kind == sList and p.items.len > 0 and (
    (p.items[0].kind == sSymbol and p.items[0].symval == "or") or
    (p.items[0].kind == sKeyword and p.items[0].kwval == "or"))
  if headIsOr:
    for i in 1 ..< p.items.len:
      parseWhereElement(p.items[i], ir, params, i - 1)  # each or-item is a branch
  elif p.kind == sList and p.items.len > 0 and
       (isOpSym(p.items[0]) or (isKeyword(p.items[0]) and p.items[0].kwval in RangeOps)):
    parsePredicate(p, ir, params, branch)
  else:
    parsePattern(p, ir, params)

proc parseDatalogQuery*(src: string): DatalogIR {.raises: [DatalogSyntaxError, ValueError], gcsafe.} =
  ## Parse a Datalog EDN query vector into the IR.
  ## Raises DatalogSyntaxError on malformed input.
  var top: SExpr
  try:
    top = readEdn(src)
  except EdnError as e:
    raise newException(DatalogSyntaxError, "datalog: EDN parse error: " & e.msg)
  if top.kind != sList or top.items.len == 0 or
     not isKeyword(top.items[0]) or top.items[0].kwval != "find":
    raise newException(DatalogSyntaxError,
      "datalog: query must start with [:find ...]")

  result = DatalogIR(rangeBounds: initTable[string, seq[seq[(string, BoundValue)]]]())

  var params = initTable[string, uint32]()
  var paramIdx: uint32 = 0

  type Section = enum secNone, secFind, secIn, secWhere
  var section: Section = secFind  # the header check guarantees items[0] = :find

  var i = 1
  while i < top.items.len:
    let el = top.items[i]
    if isKeyword(el):
      case el.kwval
      of "find": section = secFind
      of "in": section = secIn
      of "where": section = secWhere
      of "history":
        result.history = true
        section = secNone
      else:
        raise newException(DatalogSyntaxError,
          "datalog: unknown query section :" & el.kwval)
    else:
      case section
      of secFind:
        if not isVar(el):
          raise newException(DatalogSyntaxError,
            "datalog: :find takes vars like ?name, got " & $el)
        result.findVars.add FindVar(kind: fvVar, varName: varName(el))
      of secIn:
        if el.kind == sSymbol and el.symval == "$":  # the db input — ignore
          discard
        elif isVar(el):
          inc paramIdx
          params[varName(el)] = paramIdx
        else:
          raise newException(DatalogSyntaxError,
            "datalog: :in takes $ and vars like ?x, got " & $el)
      of secWhere:
        parseWhereElement(el, result, params, 0)
      of secNone:
        raise newException(DatalogSyntaxError,
          "datalog: unexpected element outside a section: " & $el)
    inc i

  if result.findVars.len == 0:
    raise newException(DatalogSyntaxError, "datalog: :find is required")
  if result.patterns.len == 0:
    raise newException(DatalogSyntaxError, "datalog: :where with at least one pattern is required")

  # Every range-predicate var must be bound by some pattern's value slot —
  # an unbound predicate var would silently filter nothing.
  block checkBound:
    for varName, branches in pairs(result.rangeBounds):
      var bound = false
      for p in result.patterns:
        if p.v.kind == dsVar and p.v.varName == varName:
          bound = true; break
      if not bound:
        raise newException(DatalogSyntaxError,
          "datalog: predicate var ?" & varName &
          " is not bound by any pattern value slot")

File: ocaml/lib/query_edn.ml
(* query_edn.ml — Datalog EDN surface → Datalog IR, mirroring
   nim_datalog/query_edn.nim.  Operates on Edn.t values from edn.ml
   (List covers both [] and ()); keywords carry the name WITHOUT the
   leading colon (edn.ml convention), so enum-style keyword values
   re-add it as ":" ^ name, exactly like the Nim reader. *)

exception Datalog_syntax_error of string

open Datalog_ast

let range_ops = [ ">"; "<"; ">="; "<="; "="; "!=" ]

let is_var_sym (e : Edn.t) : bool =
  match e with
  | Edn.Symbol s -> String.length s > 0 && s.[0] = '?'
  | _ -> false

let var_name (e : Edn.t) : string =
  match e with
  | Edn.Symbol s -> String.sub s 1 (String.length s - 1)
  | _ -> assert false

let is_blank (e : Edn.t) : bool = e = Edn.Symbol "_"

let parse_value (e : Edn.t) (params : (string * int32) list) : bound_value =
  if is_var_sym e then
    let vn = var_name e in
    match List.assoc_opt vn params with
    | Some idx -> Bv_param idx
    | None ->
      raise (Datalog_syntax_error
               ("datalog: unbound var ?" ^ vn ^ " in value position (declare it in :in)"))
  else
    match e with
    | Edn.Str s -> Bv_str s
    | Edn.Int i -> Bv_int i
    | Edn.Float f -> Bv_float f
    | Edn.Bool b -> Bv_bool b
    | Edn.Keyword kw -> Bv_str (":" ^ kw)  (* enum-style keyword value *)
    | _ -> raise (Datalog_syntax_error "datalog: unsupported value")

let parse_pattern (p : Edn.t) (ir : ir) (params : (string * int32) list) : unit =
  match p with
  | Edn.List [ e_slot; a_slot; v_slot ] ->
    let e =
      if is_var_sym e_slot then
        let vn = var_name e_slot in
        (* :in var in slot E binds at runtime — param (probe/seek) *)
        (match List.assoc_opt vn params with
         | Some idx -> Ds_const (Bv_param idx)
         | None -> Ds_var vn)
      else if is_blank e_slot then Ds_missing
      else (match e_slot with
            | Edn.Int i -> Ds_const (Bv_int i)
            | _ -> raise (Datalog_syntax_error
                            "datalog: e slot must be a var, _ or eid"))
    in
    let attr =
      match a_slot with
      | Edn.Keyword kw ->
        (* SQL-style surface normalizes dot → slash BEFORE the ns check *)
        let attr =
          if String.contains kw '.' then
            String.map (fun c -> if c = '.' then '/' else c) kw
          else kw
        in
        if not (String.contains attr '/') then
          raise (Datalog_syntax_error ("datalog: attr keyword must be namespaced: :" ^ kw));
        attr
      | _ -> raise (Datalog_syntax_error
                      "datalog: attr slot must be a keyword like :person/name")
    in
    let a = Ds_const (Bv_attr attr) in
    let v =
      if is_var_sym v_slot then
        let vn = var_name v_slot in
        (match List.assoc_opt vn params with
         | Some idx -> Ds_const (Bv_param idx)
         | None -> Ds_var vn)
      else if is_blank v_slot then Ds_missing
      else Ds_const (parse_value v_slot params)
    in
    ir.patterns <- ir.patterns @ [ { e; a; v; t = Ds_missing; added = Ds_missing } ]
  | _ -> raise (Datalog_syntax_error
                  "datalog: pattern must be a 3-element vector [e attr v]")

let pred_key (op_expr : Edn.t) : string =
  match op_expr with
  | Edn.Keyword op when List.mem op range_ops -> op
  | Edn.Keyword op ->
    raise (Datalog_syntax_error
             ("datalog: unsupported predicate op :" ^ op ^
              " (supported: > < >= <= = !=)"))
  | _ -> raise (Datalog_syntax_error
                  "datalog: predicate op must be a keyword like :>")

let is_op_sym (e : Edn.t) : bool =
  match e with Edn.Symbol s -> List.mem s range_ops | _ -> false

let add_range (ir : ir) (vn : string) (branch : int) (cond : range_cond) : unit =
  let branches = match range_bounds_lookup ir vn with Some b -> b | None -> [] in
  let arr = Array.make (max (List.length branches) (branch + 1)) [] in
  List.iteri (fun i b -> arr.(i) <- b) branches;
  arr.(branch) <- arr.(branch) @ [ cond ];
  range_bounds_set ir vn (Array.to_list arr)

let parse_predicate (p : Edn.t) (ir : ir) (params : (string * int32) list)
    (branch : int) : unit =
  let is_in head =
    match head with
    | Edn.Symbol "in" -> true
    | Edn.Keyword "in" -> true
    | _ -> false
  in
  match p with
  | Edn.List (head :: rest) when rest <> [] && is_in head ->
    (* [(in ?var v1 v2 ...)] — disjunction of equalities *)
    let var_expr = List.hd rest in
    if not (is_var_sym var_expr) then
      raise (Datalog_syntax_error "datalog: IN var must be ?var");
    let vn = var_name var_expr in
    List.iter
      (fun item -> add_range ir vn branch ("in", parse_value item params))
      (List.tl rest)
  | Edn.List [ op_expr; var_expr; val_expr ] ->
    let op =
      if is_op_sym op_expr then
        (match op_expr with Edn.Symbol s -> s | _ -> assert false)
      else pred_key op_expr
    in
    if not (is_var_sym var_expr) then
      raise (Datalog_syntax_error "datalog: predicate var must be ?var");
    let vn = var_name var_expr in
    add_range ir vn branch (op, parse_value val_expr params)
  | _ -> raise (Datalog_syntax_error
                  "datalog: predicate must be [(op var value)]")

let rec parse_where_element (p0 : Edn.t) (ir : ir) (params : (string * int32) list)
    (branch : int) : unit =
  (* The EDN reader nests predicate vectors: [(> ?v 5)] arrives as
     [[> ?v 5]].  Unwrap one level, then route. *)
  let p =
    match p0 with
    | Edn.List [ Edn.List inner ] when inner <> [] ->
      let head = List.hd inner in
      let inner_is_route =
        is_op_sym head
        || head = Edn.Symbol "or" || head = Edn.Symbol "in"
        || (match head with
            | Edn.Keyword kw -> kw = "or" || kw = "in" || List.mem kw range_ops
            | _ -> false)
      in
      if inner_is_route then Edn.List inner else p0
    | _ -> p0
  in
  let head_is_or =
    match p with
    | Edn.List (head :: _) ->
      head = Edn.Symbol "or" ||
             (match head with Edn.Keyword "or" -> true | _ -> false)
    | _ -> false
  in
  if head_is_or then
    match p with
    | Edn.List (_ :: items) ->
      List.iteri
        (fun i item -> parse_where_element item ir params i)  (* or-item = branch *)
        items
    | _ -> assert false
  else
    match p with
    | Edn.List (head :: _) when
        is_op_sym head
        || head = Edn.Symbol "in"
        || (match head with
            | Edn.Keyword kw -> kw = "in" || List.mem kw range_ops
            | _ -> false) ->
      parse_predicate p ir params branch
    | _ -> parse_pattern p ir params

type section = Sec_none | Sec_find | Sec_in | Sec_where

let parse_datalog_query (src : string) : ir * string list =
  let top =
    try Edn.read_edn src
    with Edn.Edn_error msg ->
      raise (Datalog_syntax_error ("datalog: EDN parse error: " ^ msg))
  in
  let items =
    match top with
    | Edn.List items -> items
    | _ -> raise (Datalog_syntax_error "datalog: query must start with [:find ...]")
  in
  let starts_find =
    match items with
    | Edn.Keyword "find" :: _ -> true
    | _ -> false
  in
  if items = [] || not starts_find then
    raise (Datalog_syntax_error "datalog: query must start with [:find ...]");
  let ir = empty_ir () in
  let params = ref [] in
  let param_idx = ref 0l in
  let section = ref Sec_find in
  List.iteri
    (fun i el ->
      if i = 0 then ()
      else
        match el with
        | Edn.Keyword kw ->
          (match kw with
           | "find" -> section := Sec_find
           | "in" -> section := Sec_in
           | "where" -> section := Sec_where
           | "history" -> ir.history <- true; section := Sec_none
           | other -> raise (Datalog_syntax_error
                               ("datalog: unknown query section :" ^ other)))
        | _ ->
          (match !section with
           | Sec_find ->
             if not (is_var_sym el) then
               raise (Datalog_syntax_error
                        "datalog: :find takes vars like ?name");
             ir.find_vars <- ir.find_vars @ [ Fv_var (var_name el) ]
           | Sec_in ->
             (match el with
              | Edn.Symbol "$" -> ()  (* the db input *)
              | _ when is_var_sym el ->
                param_idx := Int32.succ !param_idx;
                params := !params @ [ (var_name el, !param_idx) ]
              | _ -> raise (Datalog_syntax_error
                              "datalog: :in takes $ and vars like ?x"))
           | Sec_where -> parse_where_element el ir !params 0
           | Sec_none ->
             raise (Datalog_syntax_error
                      "datalog: unexpected element outside a section"))
    )
    items;
  if ir.find_vars = [] then
    raise (Datalog_syntax_error "datalog: :find is required");
  if ir.patterns = [] then
    raise (Datalog_syntax_error
             "datalog: :where with at least one pattern is required");
  (* every range-predicate var must be bound by some pattern value slot *)
  List.iter
    (fun (vn, _) ->
      let bound =
        List.exists
          (fun (p : pattern) -> match p.v with Ds_var name -> name = vn | _ -> false)
          ir.patterns
      in
      if not bound then
        raise (Datalog_syntax_error
                 ("datalog: predicate var ?" ^ vn ^
                  " is not bound by any pattern value slot")))
    ir.range_bounds;
  let find_vars =
    List.map
      (function Fv_var n -> n | Fv_const (n, _) -> n)
      ir.find_vars
  in
  (ir, find_vars)
